contracts/0.8.9/WithdrawalQueue.sol

```diff
@@ -277,21 +279,79 @@ contract WithdrawalQueue is AccessControlEnumerable, WithdrawalQueueBase {
         return _requestWithdrawalWstETH(_amountOfWstETH, _recipient);
     }
 
-    /// @notice Request withdrawal of the provided token amount in a batch
-    /// NB: Always reverts
-    function requestWithdrawalBatch(uint256[] calldata, address[] calldata)
+    struct WithdrawalRequestInput {
+        address token;
+        uint256 amount;
+        address recipient;
+    }
+
+    /// @notice Request the sequence of withdrawals according to passed `withdrawalRequestInputs` data
+    /// @param _withdrawalRequestInputs an array of `WithdrawalRequestInput` data. The standalone withdrawal request will
+    ///  be created for each item in the passed list. If `WithdrawalRequestInput.recipient` is set to `address(0)`,
+    ///  `msg.sender` will be used as recipient.
+    /// @return requestIds an array of the created withdrawal requests
+    function requestWithdrawals(WithdrawalRequestInput[] memory _withdrawalRequestInputs)
         external
-        view
         whenResumed
-        returns (uint256[] memory)
+        returns (uint256[] memory requestIds)
     {
-        revert Unimplemented();
+        requestIds = new uint256[](_withdrawalRequestInputs.length);
+        for (uint256 i = 0; i < _withdrawalRequestInputs.length; ++i) {
+            requestIds[i] = _enqueueWithdrawalRequest(msg.sender, _withdrawalRequestInputs[i]);
+        }
+    }
+
+    struct Permit {
+        uint8 v;
+        bytes32 r;
+        bytes32 s;
+        address owner;
+        uint256 deadline;
+    }
+
+    /// @notice Request the sequence of withdrawals according to passed `withdrawalRequestInputs` data using EIP-2612 Permit
+    /// @param _withdrawalRequestInputs an array of `WithdrawalRequestInput` data. The standalone withdrawal request will
+    ///  be created for each item in the passed list. If `WithdrawalRequestInput.recipient` is set to `address(0)`,
+    ///  `msg.sender` will be used as recipient.
+    /// @return requestIds an array of the created withdrawal requests
+    function requestWithdrawalsWithPermit(
+        WithdrawalRequestInput[] memory _withdrawalRequestInputs,
+        Permit[] memory _permits
```

Comment:
not sure that we need an array of permits, tbh
maybe just one for the whole sum?

Reply:
Fixed. Made batch methods more strict:
- standalone methods for stETH and wstETH
- one permit expected for the whole withdrawal amount